lib/cfg.c

```diff
@@ -535,16 +535,16 @@ cfg_read_config(GlobalConfig *self, const gchar *fname, gboolean syntax_only, gc
   if ((cfg_file = fopen(fname, "r")) != NULL)
     {
       CfgLexer *lexer;
-      GString *preprocess_output = g_string_sized_new(8192);
+      self->preprocess_config = g_string_sized_new(8192);
 
-      lexer = cfg_lexer_new(self, cfg_file, fname, preprocess_output);
+      lexer = cfg_lexer_new(self, cfg_file, fname, self->preprocess_config);
       res = cfg_run_parser(self, lexer, &main_parser, (gpointer *) &self, NULL);
       fclose(cfg_file);
       if (preprocess_into)
         {
-          cfg_dump_processed_config(preprocess_output, preprocess_into);
+          cfg_dump_processed_config(self->preprocess_config, preprocess_into);
         }
-      g_string_free(preprocess_output, TRUE);
```

Comment:
My question is, is it safe to move the g_string_free to cfg_free?

In the `main_loop_read_and_init_config` I see the only place when `cfg_free` is not called in the failure branch, that might leak this value. Can we add a `cfg_free` there? Otherwise I think g_string_free method could be moved into `cfg_free`.

Reply:
I do not think there is much choice about using `cfg_free`, the `GlobalConfig` own's that string, that should be freed by `cfg_free`.

On the other hand your findings - after just a glimpse - looks valid, but that is a general issue. I do not want to mix that with this PR.

Reply:
Okok.

Reply:
Correct me if I'm wrong but I think that possible leak you mentioned does not have a high severity as it would only happen in an error case, after which syslog-ng is terminated.
This all happens on early startup, it's not affecting production environment.

Of course an error-handling is missing and we could build something else on `main_loop_read_and_init_config` too with the possibility to leak memory.

Although it is a nice catch! :+1: